### ISSUE TITLE
[FR] Allow inequality conditions in `block_by`

### ISSUE THREAD

**Issue by etiennebacher** [beniaminogreen/zoomerjoin#100]
**Is your feature request related to a problem? Please describe.**
It is nice to be able to use `block_by` to filter out some comparisons before computing the string similarity. Currently, it is limited to equality conditions (e.g rows must have the same year to be considered for matching). I have a setting in which I don't want to compare rows if `year.y` is larger than `year.x`, i.e I'd like to block matching by the condition `year.x > year.y`.

I don't know how hard that would be to implement. I could track the usage of `block_by` (renamed `salt` in the internals) until the method `new()` for the `Singleset` struct in Rust, but I don't know how that works exactly.

**Describe the solution you'd like**
Not sure of the syntax this should take. It could be something similar to `dplyr::join_by()`, so that one could do `block_by = block_by(year.x > year.y)`. Also, that only works if the variable has a different name in each dataset. 

**Describe alternatives you've considered**
Currently I don't use `block_by`. Instead I match on the full data and filter out by my condition afterwards.

**Additional context**
/

**Comment by beniaminogreen** [comment]
This is a very interesting feature request, but I think that it could be tough to implement. It might also be too far outside what I see as the main goals for the package, namely to accelerate big-data joins with locality sensitive hashing techniques. 

I am ambivalent about the current blocking functionality in the package and if I could go back in time, I might not have included it as it creates extra code to maintain. If it is useful to people (and it sounds like some are using it), I would also be happy to change my mind about it. 

The reason I originally included the blocking is that it's easy to implement / add on to the existing Locality Sensitive Hashing algorithms. Internally the blocking variables are called "salts" because they are literally used to salt hashes. This ensures that only units with the same salting / blocking variables hash collisions, which is a prerequisite for being compared as a potential match. 

If we want to add an inequality constraint it seems much tougher to do this within the locality sensitive hashing framework. If we can find away to do it, I would love to have a try at implementing it.  I can ask a professor I know about this possibility. 

If we can't find a way to implement the inequality blocking as part of the locality sensitive hashing, we would likely have to add on the inequality as a filter step once the matching rows have been returned from the LSH algorithm. In this case, I think we shouldn't call it blocking as much as we should include it as part of the `by` argument (i.e. by = by(name_1 == name_2, year_1 > year_2)).  I think that I would be very slightly less enthusiastic about this approach, but I would also be happy to implement it if you think there is sufficient demand for this kind of functionality. 

Best, 
Ben

**Comment by etiennebacher** [comment]
Thanks for the detailed answer! I have to think more about this (also curious to know what the prof you mention would say).

> If we can't find a way to implement the inequality blocking as part of the locality sensitive hashing, we would likely have to add on the inequality as a filter step once the matching rows have been returned from the LSH algorithm.

I don't think this is worth implementing in the package. Blocking on inequality conditions would be a nice feature of the package IMO, because it reduces the work that the LSH has to do, but filtering afterwards based on some condition is a simple operation and is not related to LSH anymore.

**Comment by beniaminogreen** [comment]
Gotcha. Sounds like we are totally of the same mind about this. I will continue searching to see if there's a way to integrate the inequality constraints with the LSH and will keep you posted if I find anything. Until then, I won't try and add any filters to the package.

**Comment by etiennebacher** [comment]
Thank you for the potential solution, I must say I don't have a usecase for this anymore so I won't try to benchmark the two approaches.

I'm wondering it using `polars` as a backend would make things faster. There is a mechanism to write extensions in Rust that can directly be applied on `polars` DataFrames and LazyFrames (so far this is only available in Python Polars but I hope those extensions can be soon possible in r-polars).

**Comment by etiennebacher** [comment]
> My only concern right now is that we are very close to the upper limit (5MB) that CRAN imposes on vendored dependencies, and I worry that adding the polars crate could push us over the edge

True, this could be an issue, although I'm not sure how many of `polars` several "subcrates" would be needed for this. Let's see in a few weeks / months if we have a working mechanism for extensions in r-polars.

> Do you have any experience distributing packages with large vendored dependencies on CRAN? Does the 5MB limit pose a serious constraint in practice, or are there ways to work around it?

I personally don't, but I know some people who have packages > 5MB on CRAN. I think you have to argue on submission why this is reasonable (provided that you don't go over 20MB for instance). I also know there were some discussion to download crates from a "reliable source" (according to CRAN) and I think that Zenodo was mentioned as such a source. I think the discussion is on the R-devel list of those last months.
